ssl/statem/statem_lib.c

```diff
@@ -1007,6 +1007,28 @@ int ssl_choose_client_version(SSL *s, int version)
 }
 
 /*
+ * ssl_verify_client_session_version - Verify that provided SSL_SESSION matches
+ * the protocol negotiated with server.
+ *
+ * @s: client SSL handle.
+ *
+ * Returns 1 when it matches, 0 otherwise
+ */
+int ssl_verify_client_session_version(SSL *s)
+{
+  /* Empty session and no EAP-FAST */
+  if (s->session->session_id_length == 0 && !s->session->tlsext_tick) {
```

Comment:
>Oh, I was thinking it should go slightly higher, near s->session->session_id_length = session_id_len;. This way you fix ssl_version in the fresh session case, but don't violate thread-safety in reuse.

@davidben like this? Added assignment in two places.

>The one this thread is attached to. if (s->session->session_id_length == 0 && !s->session->tlsext_tick) {. Or is that for something else? I think this function really just wants to be a version check.

@davidben removed the branch, and inlined the function.